### ISSUE TITLE
--warn-return-any should not complain on NotImplemented when using abc

### ISSUE THREAD

**Issue by drtyrsa** [python/mypy#4534]
I'm using `abc.abstractmethod`:
```python
@abc.abstractmethod
def encode(self) -> bytes:
    return NotImplemented
```

And I get a warning: `warning: Returning Any from function declared to return "bytes"`

I don't think something is wrong with the code, it's totally normal way to declare and describe an interface and mypy shouldn't complain about it.

**Comment by ilevkivskyi** [comment]
Why do you think this is a standard pattern? I always thought that `NotImplemented` is reserved for reversed operators (like `__radd__` vs `__add__`) and few other special methods (like `__subclasshook__`). I think the standard way of doing what you want is:
```python
@abc.abstractmethod
def encode(self) -> bytes:
    raise NotImplementedError("Not a valid encoder -- implement `encode` before use")
```
and I hope mypy doesn't complain about this (if yes, then _this_ should be fixed).

**Comment by gvanrossum** [comment]
You have a point there. I wonder if the correct type for `NotImplemented`
shouldn't be `object` instead of `Any`. Can you try this (it's a one-line
change in typeshed) and submit a PR?

**Comment by ilevkivskyi** [comment]
@gvanrossum It is not possible because `NotImplemented` is actually used in many `__op__` methods that return non-`Any` types. All such cases will be flagged as error by mypy if we type `NotImplemented` as `object`.

**Comment by ilevkivskyi** [comment]
A simplest example is `__eq__`, it is required to return `bool`, and therefore a standard pattern like:
```python 
def __eq__(self, other: object) -> bool:
    if isinstance(other, type(self)):
        ...
    return NotImplemented
```
will be flagged as an error.

**Comment by gvanrossum** [comment]
Right. Maybe `--disallow-any-expr` can be hacked to ignore this case?

**Comment by ilevkivskyi** [comment]
> Maybe `--disallow-any-expr` can be hacked to ignore this case?

Yes, this is easy, but would not be my priority TBH, if OP wants to make a PR then OK.

**Comment by drtyrsa** [comment]
@ilevkivskyi I've reread the docs and you are right, I shouldn't use `NotImplemented` here. I'll make a PR about `return NotImplemented` case.

Strangely abc docs doesn't say anything about what should a body of an abstract method look like. PEP 3119 doesn't either (and has an example with `pass`).

I guess I should raise an exception here, thank you.
